src/box/box.cc

```diff
@@ -757,23 +761,98 @@ box_check_say(void)
 		say_free_syslog_opts(&opts);
 	}
 
-	const char *log_format = cfg_gets("log_format");
-	enum say_format format = say_format_by_name(log_format);
+	enum say_format format = say_format_by_name(format_str);
 	if (format == say_format_MAX)
 		tnt_raise(ClientError, ER_CFG, "log_format",
 			 "expected 'plain' or 'json'");
 	if (type == SAY_LOGGER_SYSLOG && format == SF_JSON) {
 		tnt_raise(ClientError, ER_CFG, "log_format",
 			  "'json' can't be used with syslog logger");
 	}
-	int log_nonblock = cfg_getb("log_nonblock");
-	if (log_nonblock == 1 &&
+	if (nonblock == 1 &&
 	    (type == SAY_LOGGER_FILE || type == SAY_LOGGER_STDERR)) {
 		tnt_raise(ClientError, ER_CFG, "log_nonblock",
 			  "the option is incompatible with file/stderr logger");
 	}
 }
 
+/**
+ * Historically values lesser then 0 or greater then say_level_MAX are
+ * allowed if given by number.
+ */
+int
+box_get_log_level(int *level)
+{
+	if (cfg_isnumber("log_level")) {
+		*level = cfg_geti("log_level");
+	} else {
+		const char *str_level = cfg_gets("log_level");
+		*level = strindex(say_level_strs, str_level, say_level_MAX);
+		if (*level >= say_level_MAX)
+			return -1;
+	}
+	return 0;
+}
+
+void
+box_log_cfg(struct lua_State *L)
+{
+	lua_pushstring(L, "log");
+	lua_gettable(L, 1);
+	const char *log = lua_tostring(L, -1);
+
+	lua_pushstring(L, "format");
+	lua_gettable(L, 1);
+	const char *format = lua_tostring(L, -1);
+
+	int nonblock;
+	lua_pushstring(L, "nonblock");
+	lua_gettable(L, 1);
+	/* nil is a special value for logger, not false */
+	if (lua_isnil(L, -1))
+		nonblock = -1;
+	else
+		nonblock = lua_toboolean(L, -1);
+
+	int level;
+	lua_pushstring(L, "level");
+	lua_gettable(L, 1);
+	if (lua_isnumber(L, -1)) {
+		level = lua_tointeger(L, -1);
+	} else {
+		const char *str_level = lua_tostring(L, -1);
+		level = strindex(say_level_strs, str_level, say_level_MAX);
+		if (level >= say_level_MAX)
+			tnt_raise(ClientError, ER_CFG, "log_level",
+				  "invalid value");
+	}
+
+	say_cfg_check(log, format, nonblock);
+
+	/*
+	 * Using log setters before log.cfg or box.cfg should not be
+	 * treated as configuring.
+	 */
+	lua_pushstring(L, "__setter");
```

Comment:
`__setter` pattern is confusing. Please split into two functions instead.

Reply:
Fixed.